mod_zip does not compile with nginx 1.23.0+

The new nginx mainline 1.23.0 changed the way some headers are handled ( see https://nginx.org/en/CHANGES and https://hg.nginx.org/nginx/rev/ef6a3a99a81a ).

Looks like `mod_zip` needs some smaller changes to be compatible with nginx 1.23.0.

Here ist a build log of mod_zip 1.2.0 (but current `master` should be the same)

```
#11 12.75 make[2]: Entering directory '/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0/debian/build-nginx'
#11 12.75 cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O2 -fdebug-prefix-map=/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
#11 12.75 	-o objs/addon/zip-1.23.0/ngx_http_zip_module.o \
#11 12.75 	extra/zip-1.23.0/ngx_http_zip_module.c
#11 12.75 cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O2 -fdebug-prefix-map=/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
#11 12.75 	-o objs/addon/zip-1.23.0/ngx_http_zip_parsers.o \
#11 12.75 	extra/zip-1.23.0/ngx_http_zip_parsers.c
#11 12.75 cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O2 -fdebug-prefix-map=/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
#11 12.75 	-o objs/addon/zip-1.23.0/ngx_http_zip_file.o \
#11 12.75 	extra/zip-1.23.0/ngx_http_zip_file.c
#11 12.75 cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O2 -fdebug-prefix-map=/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
#11 12.75 	-o objs/addon/zip-1.23.0/ngx_http_zip_headers.o \
#11 12.75 	extra/zip-1.23.0/ngx_http_zip_headers.c
#11 12.75 cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -g -O2 -fdebug-prefix-map=/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
#11 12.75 	-o objs/ngx_http_zip_module_modules.o \
#11 12.75 	objs/ngx_http_zip_module_modules.c
#11 12.95 extra/zip-1.23.0/ngx_http_zip_headers.c: In function 'ngx_http_zip_add_cache_control':
#11 12.95 extra/zip-1.23.0/ngx_http_zip_headers.c:13:39: error: 'r->headers_out.cache_control' is a pointer; did you mean to use '->'?
#11 12.95    13 |     ccp = r->headers_out.cache_control.elts;
#11 12.95       |                                       ^
#11 12.95       |                                       ->
#11 12.95 extra/zip-1.23.0/ngx_http_zip_headers.c:16:28: error: passing argument 1 of 'ngx_array_init' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95    16 |         if (ngx_array_init(&r->headers_out.cache_control, r->pool,
#11 12.95       |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                            |
#11 12.95       |                            ngx_table_elt_t ** {aka struct ngx_table_elt_s **}
#11 12.95 In file included from src/core/ngx_core.h:65,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:4,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_headers.c:1:
#11 12.95 src/core/ngx_array.h:32:29: note: expected 'ngx_array_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_table_elt_t **' {aka 'struct ngx_table_elt_s **'}
#11 12.95    32 | ngx_array_init(ngx_array_t *array, ngx_pool_t *pool, ngx_uint_t n, size_t size)
#11 12.95       |                ~~~~~~~~~~~~~^~~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_headers.c:23:30: error: passing argument 1 of 'ngx_array_push' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95    23 |         ccp = ngx_array_push(&r->headers_out.cache_control);
#11 12.95       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                              |
#11 12.95       |                              ngx_table_elt_t ** {aka struct ngx_table_elt_s **}
#11 12.95 In file included from src/core/ngx_core.h:65,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:4,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_headers.c:1:
#11 12.95 src/core/ngx_array.h:27:7: note: expected 'ngx_array_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_table_elt_t **' {aka 'struct ngx_table_elt_s **'}
#11 12.95    27 | void *ngx_array_push(ngx_array_t *a);
#11 12.95       |       ^~~~~~~~~~~~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_headers.c:39:53: error: 'r->headers_out.cache_control' is a pointer; did you mean to use '->'?
#11 12.95    39 |         for (i = 1; i < r->headers_out.cache_control.nelts; i++) {
#11 12.95       |                                                     ^
#11 12.95       |                                                     ->
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c: In function 'ngx_http_zip_generate_pieces':
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:254:67: error: passing argument 1 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   254 |         variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_name_separator,
#11 12.95       |                                                                   ^~
#11 12.95       |                                                                   |
#11 12.95       |                                                                   ngx_http_variable_value_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:60:64: note: expected 'ngx_http_request_t *' {aka 'struct ngx_http_request_s *'} but argument is of type 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'}
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |                                            ~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:254:71: error: passing argument 2 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   254 |         variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_name_separator,
#11 12.95       |                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                                                       |
#11 12.95       |                                                                       ngx_str_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:32: note: expected 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_str_t *' {aka 'struct <anonymous> *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:255:17: error: passing argument 3 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   255 |                 &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
#11 12.95       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                 |
#11 12.95       |                 ngx_list_part_t * {aka struct ngx_list_part_s *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:46: note: expected 'ngx_str_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                   ~~~~~~~~~~~^~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:255:56: error: passing argument 4 of 'ngx_http_variable_unknown_header' makes pointer from integer without a cast [-Werror=int-conversion]
#11 12.95   255 |                 &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
#11 12.95       |                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                                        |
#11 12.95       |                                                        long unsigned int
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:68: note: expected 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'} but argument is of type 'long unsigned int'
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                                   ~~~~~~~~~~~~~~~~~^~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:254:34: error: too few arguments to function 'ngx_http_variable_unknown_header'
#11 12.95   254 |         variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_name_separator,
#11 12.95       |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:60:11: note: declared here
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:268:71: error: passing argument 1 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   268 |             variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_charset_name,
#11 12.95       |                                                                       ^~
#11 12.95       |                                                                       |
#11 12.95       |                                                                       ngx_http_variable_value_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:60:64: note: expected 'ngx_http_request_t *' {aka 'struct ngx_http_request_s *'} but argument is of type 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'}
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |                                            ~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:268:75: error: passing argument 2 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   268 |             variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_charset_name,
#11 12.95       |                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                                                           |
#11 12.95       |                                                                           ngx_str_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:32: note: expected 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_str_t *' {aka 'struct <anonymous> *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:269:21: error: passing argument 3 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   269 |                     &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
#11 12.95       |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                     |
#11 12.95       |                     ngx_list_part_t * {aka struct ngx_list_part_s *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:46: note: expected 'ngx_str_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                   ~~~~~~~~~~~^~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c: In function 'ngx_http_zip_main_request_header_filter':
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:269:60: error: passing argument 4 of 'ngx_http_variable_unknown_header' makes pointer from integer without a cast [-Werror=int-conversion]
#11 12.95   269 |                     &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
#11 12.95       |                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                                            |
#11 12.95       |                                                            long unsigned int
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:202:67: error: passing argument 1 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   202 |         variable_header_status = ngx_http_variable_unknown_header(vv,
#11 12.95       |                                                                   ^~
#11 12.95       |                                                                   |
#11 12.95       |                                                                   ngx_http_variable_value_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:61:68: note: expected 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'} but argument is of type 'long unsigned int'
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                                   ~~~~~~~~~~~~~~~~~^~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_file.c:268:38: error: too few arguments to function 'ngx_http_variable_unknown_header'
#11 12.95   268 |             variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_charset_name,
#11 12.95       |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_file.c:2:
#11 12.95 src/http/ngx_http_variables.h:60:11: note: declared here
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:60:64: note: expected 'ngx_http_request_t *' {aka 'struct ngx_http_request_s *'} but argument is of type 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'}
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |                                            ~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:203:17: error: passing argument 2 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   203 |                 &ngx_http_zip_header_variable_name,
#11 12.95       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                 |
#11 12.95       |                 ngx_str_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:32: note: expected 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_str_t *' {aka 'struct <anonymous> *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:204:17: error: passing argument 3 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   204 |                 &r->upstream->headers_in.headers.part, sizeof("upstream_http_") - 1);
#11 12.95       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                 |
#11 12.95       |                 ngx_list_part_t * {aka struct ngx_list_part_s *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:46: note: expected 'ngx_str_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                   ~~~~~~~~~~~^~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:204:56: error: passing argument 4 of 'ngx_http_variable_unknown_header' makes pointer from integer without a cast [-Werror=int-conversion]
#11 12.95   204 |                 &r->upstream->headers_in.headers.part, sizeof("upstream_http_") - 1);
#11 12.95       |                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                                        |
#11 12.95       |                                                        long unsigned int
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:68: note: expected 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'} but argument is of type 'long unsigned int'
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                                   ~~~~~~~~~~~~~~~~~^~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:202:34: error: too few arguments to function 'ngx_http_variable_unknown_header'
#11 12.95   202 |         variable_header_status = ngx_http_variable_unknown_header(vv,
#11 12.95       |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:60:11: note: declared here
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:206:67: error: passing argument 1 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   206 |         variable_header_status = ngx_http_variable_unknown_header(vv,
#11 12.95       |                                                                   ^~
#11 12.95       |                                                                   |
#11 12.95       |                                                                   ngx_http_variable_value_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:60:64: note: expected 'ngx_http_request_t *' {aka 'struct ngx_http_request_s *'} but argument is of type 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'}
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |                                            ~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:207:17: error: passing argument 2 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   207 |                 &ngx_http_zip_header_variable_name,
#11 12.95       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                 |
#11 12.95       |                 ngx_str_t * {aka struct <anonymous> *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:32: note: expected 'ngx_http_variable_value_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_str_t *' {aka 'struct <anonymous> *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:208:17: error: passing argument 3 of 'ngx_http_variable_unknown_header' from incompatible pointer type [-Werror=incompatible-pointer-types]
#11 12.95   208 |                 &r->headers_out.headers.part, sizeof("upstream_http_") - 1);
#11 12.95       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                 |
#11 12.95       |                 ngx_list_part_t * {aka struct ngx_list_part_s *}
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:46: note: expected 'ngx_str_t *' {aka 'struct <anonymous> *'} but argument is of type 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'}
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                   ~~~~~~~~~~~^~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:208:47: error: passing argument 4 of 'ngx_http_variable_unknown_header' makes pointer from integer without a cast [-Werror=int-conversion]
#11 12.95   208 |                 &r->headers_out.headers.part, sizeof("upstream_http_") - 1);
#11 12.95       |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95       |                                               |
#11 12.95       |                                               long unsigned int
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:61:68: note: expected 'ngx_list_part_t *' {aka 'struct ngx_list_part_s *'} but argument is of type 'long unsigned int'
#11 12.95    61 |     ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
#11 12.95       |                                                   ~~~~~~~~~~~~~~~~~^~~~
#11 12.95 extra/zip-1.23.0/ngx_http_zip_module.c:206:34: error: too few arguments to function 'ngx_http_variable_unknown_header'
#11 12.95   206 |         variable_header_status = ngx_http_variable_unknown_header(vv,
#11 12.95       |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.95 In file included from src/http/ngx_http.h:30,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.h:5,
#11 12.95                  from extra/zip-1.23.0/ngx_http_zip_module.c:9:
#11 12.95 src/http/ngx_http_variables.h:60:11: note: declared here
#11 12.95    60 | ngx_int_t ngx_http_variable_unknown_header(ngx_http_request_t *r,
#11 12.95       |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#11 12.96 cc1: all warnings being treated as errors
#11 12.96 cc1: all warnings being treated as errors
#11 12.96 cc1: all warnings being treated as errors
#11 12.96 make[2]: *** [objs/Makefile:1881: objs/addon/zip-1.23.0/ngx_http_zip_headers.o] Error 1
#11 12.96 make[2]: *** Waiting for unfinished jobs....
#11 12.97 make[2]: *** [objs/Makefile:1860: objs/addon/zip-1.23.0/ngx_http_zip_module.o] Error 1
#11 12.97 make[2]: *** [objs/Makefile:1874: objs/addon/zip-1.23.0/ngx_http_zip_file.o] Error 1
#11 13.06 make[2]: Leaving directory '/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0/debian/build-nginx'
#11 13.06 make[1]: *** [Makefile:16: modules] Error 2
#11 13.06 make[1]: Leaving directory '/build/pkg-oss/debian/debuild-module-zip/nginx-1.23.0/debian/build-nginx'
#11 13.06 make: *** [debian/rules:72: build-arch.nginx] Error 2
#11 13.06 dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,7 +7,7 @@ jobs:
       fail-fast: false
       matrix:
         compiler: [ gcc, clang ]
-        version: [ 1.19.10, 1.20.2, 1.21.4 ]
+        version: [ 1.19.10, 1.20.2, 1.21.4, 1.22.0, 1.23.0 ]
     steps:
       - name: Install carton
         run: sudo apt install carton

--- a/config
+++ b/config
@@ -36,3 +36,15 @@ if [ $ngx_found = yes ]; then
     CORE_LIBS="$CORE_LIBS $ngx_feature_libs"
     ngx_found=no
 fi
+
+ngx_feature="multi-value headers are linked lists (nginx 1.23.0+)"
+ngx_feature_name="NGX_ZIP_MULTI_HEADERS_LINKED_LISTS"
+ngx_feature_run=yes
+ngx_feature_incs="#include \"nginx.h\""
+ngx_feature_path="src/core"
+ngx_feature_test="
+#if nginx_version < 1023000
+#error multi headers are no linked lists
+#endif
+"
+. auto/feature

--- a/ngx_http_zip_file.c
+++ b/ngx_http_zip_file.c
@@ -3,6 +3,7 @@
 #include "ngx_http_zip_file.h"
 #include "ngx_http_zip_file_format.h"
 #include "ngx_http_zip_endian.h"
+#include "ngx_http_zip_headers.h"
 
 #ifdef NGX_ZIP_HAVE_ICONV
 #include <iconv.h>
@@ -252,7 +253,7 @@ ngx_http_zip_generate_pieces(ngx_http_request_t *r, ngx_http_zip_ctx_t *ctx)
     // because UFT-8 flag (zip_utf8_flag) is set default for templates.
     ngx_int_t variable_header_status = NGX_OK;
     if (r->upstream) {
-        variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_name_separator,
+        variable_header_status = ngx_http_zip_variable_unknown_header(r, vv, &ngx_http_zip_header_name_separator,
                 &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
     } else {
         vv->not_found = 1;
@@ -266,7 +267,7 @@ ngx_http_zip_generate_pieces(ngx_http_request_t *r, ngx_http_zip_ctx_t *ctx)
 #ifdef NGX_ZIP_HAVE_ICONV
         variable_header_status = NGX_OK;
         if (r->upstream) {
-            variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_charset_name,
+            variable_header_status = ngx_http_zip_variable_unknown_header(r, vv, &ngx_http_zip_header_charset_name,
                     &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
         } else {
             vv->not_found = 1;
@@ -431,7 +432,7 @@ ngx_http_zip_generate_pieces(ngx_http_request_t *r, ngx_http_zip_ctx_t *ctx)
 
     // Collect names of original request's header fields that
     // have to be present in each of the issued sub-requests.
-    variable_header_status = ngx_http_variable_unknown_header(vv, &ngx_http_zip_header_name_pass_headers,
+    variable_header_status = ngx_http_zip_variable_unknown_header(r, vv, &ngx_http_zip_header_name_pass_headers,
             &r->upstream->headers_in.headers.part, sizeof("upstream_http_")-1);
 
     if (variable_header_status == NGX_OK && !vv->not_found) {

--- a/ngx_http_zip_headers.c
+++ b/ngx_http_zip_headers.c
@@ -21,6 +21,36 @@ ngx_http_zip_find_key_in_set(ngx_str_t *key, ngx_array_t *set)
 ngx_int_t 
 ngx_http_zip_add_cache_control(ngx_http_request_t *r)
 {
+#ifdef NGX_ZIP_MULTI_HEADERS_LINKED_LISTS
+    ngx_table_elt_t            *cc;
+
+    /* convoluted way of adding Cache-Control: max-age=0 */
+    /* The header is necessary so IE doesn't barf */
+    cc = r->headers_out.cache_control;
+
+    if (cc == NULL) {
+        cc = ngx_list_push(&r->headers_out.headers);
+        if (cc == NULL) {
+            return NGX_ERROR;
+        }
+
+        r->headers_out.cache_control = cc;
+        cc->next = NULL;
+        cc->hash = 1;
+        ngx_str_set(&cc->key, "Cache-Control");
+    } else {
+         for (cc = cc->next; cc; cc = cc->next) {
+            cc->hash = 0;
+         }
+
+         cc = r->headers_out.cache_control;
+         cc->next = NULL;
+    }
+
+    ngx_str_set(&cc->value, "max-age=0");
+
+    return NGX_OK;
+#else
     ngx_table_elt_t           **ccp, *cc;
     ngx_uint_t                  i;
 
@@ -62,6 +92,7 @@ ngx_http_zip_add_cache_control(ngx_http_request_t *r)
     ngx_str_set(&cc->value, "max-age=0");
 
     return NGX_OK;
+#endif
 }
 
 ngx_int_t 
@@ -284,4 +315,17 @@ ngx_http_zip_init_subrequest_headers(ngx_http_request_t *r, ngx_http_zip_ctx_t *
     }
 
     return NGX_OK;
+}
+
+
+ngx_int_t
+ngx_http_zip_variable_unknown_header(ngx_http_request_t *r,
+                                 ngx_http_variable_value_t *v, ngx_str_t *var,
+                                 ngx_list_part_t *part, size_t prefix)
+{
+#ifdef NGX_ZIP_MULTI_HEADERS_LINKED_LISTS
+    return ngx_http_variable_unknown_header(r, v, var, part, prefix);
+#else
+    return ngx_http_variable_unknown_header(v, var, part, prefix);
+#endif
 }

--- a/ngx_http_zip_headers.h
+++ b/ngx_http_zip_headers.h
@@ -13,3 +13,7 @@ ngx_int_t ngx_http_zip_init_multipart_range(ngx_http_request_t *r,
 ngx_int_t ngx_http_zip_init_subrequest_headers(ngx_http_request_t *r, ngx_http_zip_ctx_t *ctx,
         ngx_http_request_t *sr, ngx_http_zip_range_t *piece_range,
         ngx_http_zip_range_t *req_range);
+
+ngx_int_t ngx_http_zip_variable_unknown_header(ngx_http_request_t *r,
+                                           ngx_http_variable_value_t *v, ngx_str_t *var, ngx_list_part_t *part,
+                                           size_t prefix);

--- a/ngx_http_zip_module.c
+++ b/ngx_http_zip_module.c
@@ -201,13 +201,13 @@ ngx_http_zip_main_request_header_filter(ngx_http_request_t *r)
     /* Look for X-Archive-Files */
     ngx_int_t variable_header_status = NGX_OK;
     if (r->upstream) {
-        variable_header_status = ngx_http_variable_unknown_header(vv,
+        variable_header_status = ngx_http_zip_variable_unknown_header(r, vv,
                 &ngx_http_zip_header_variable_name,
-                &r->upstream->headers_in.headers.part, sizeof("upstream_http_") - 1); 
+                &r->upstream->headers_in.headers.part, sizeof("upstream_http_") - 1);
     } else if (r->headers_out.status == NGX_HTTP_OK) {
-        variable_header_status = ngx_http_variable_unknown_header(vv,
+        variable_header_status = ngx_http_zip_variable_unknown_header(r, vv,
                 &ngx_http_zip_header_variable_name,
-                &r->headers_out.headers.part, sizeof("upstream_http_") - 1); 
+                &r->headers_out.headers.part, sizeof("upstream_http_") - 1);
     } else {
         vv->not_found = 1;
     }

